### PR TITLE
ScaLAPACKMatrix: change setting of state variable at various places

### PR DESCRIPTION
**Pull request by BenBrands**

### CODE REVIEW

File: source/lac/scalapack.cc [hunk 1]
```diff
@@ -472,6 +472,7 @@ template <typename NumberType>
 void ScaLAPACKMatrix<NumberType>::copy_transposed(const ScaLAPACKMatrix<NumberType> &B)
 {
   add(B,0,1,true);
+  state = B.state;
```
Comment (davydden) [inline source/lac/scalapack.cc]: i would say move that before `add()` and inside `add()` check that `state==B.state`?

Comment (BenBrands) [reply source/lac/scalapack.cc]: I definitely agree with moving the reassignment.
Matrix B could have, in my opinion, any state and therefore a check is not needed.
The user should be aware which operations are performed.

**Comment by davydden** [comment]
> Matrix B could have, in my opinion, any state and therefore a check is not needed.
The user should be aware which operations are performed.

i guess the question we need to decide is what happens if matrix `A` (this) is in a non-matrix state (say `LU`) and user calls `copy_transposed()` or `add()`? That's why I thought that `copy_transposed()` should reset whatever this object has as a state and then `add()` should at least make sure the states are equal. But maybe even check that the state is `matrix`, as it probably does not make sense to add, say, `Cholesky` factorizations. Same with `eigenvalues` and other states we have.

So to be on the defensive side and catch such bugs, i would still add Asserts.

**Comment by BenBrands** [comment]
I think we agree on setting the state for `copy_transposed()!`.
The `state` variable is meant to ensure that a matrix is in a specific state when calling a certain `ScaLAPACK` routine/wrapper, e.g. for inversion via `ppotri` the state has to be `cholesky`. Any manipulations like `addition` or `multiplication` will reset the state to `matrix`, which is the most general. Consequently, that manipulations will render a call to routines demanding a specific state to throw an exception.
In general, it might be to restrictive to require an equal state for `addition`. I would leave it to the user to call the proper operations. After all, a `ScaLAPACKMatrix` object is just a matrix and its content should be changeable to the highest degree.

**Pull request merged**